### ISSUE TITLE
test_clone.test_clone_unborn_head_sub fails on crippled FS

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4727]
Happened with gh-4708 running on an exFAT SSD.

This is particularly bad, as it affects `core` code.
```
======================================================================
ERROR: datalad.core.distributed.tests.test_clone.test_clone_unborn_head_sub
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mih/env/datalad3-dev/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/mih/hacking/datalad/git/datalad/tests/utils.py", line 715, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/mih/hacking/datalad/git/datalad/core/distributed/tests/test_clone.py", line 1126, in test_clone_unborn_head_sub
    "sub", result_xfm="datasets", return_type="item-or-list")
  File "/home/mih/hacking/datalad/git/datalad/distribution/dataset.py", line 498, in apply_func
    return f(**kwargs)
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 494, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 482, in return_func
    results = list(results)
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 413, in generator_func
    allkwargs):
  File "/home/mih/hacking/datalad/git/datalad/interface/utils.py", line 552, in _process_results
    for res in results:
  File "/home/mih/hacking/datalad/git/datalad/distribution/get.py", line 801, in __call__
    description):
  File "/home/mih/hacking/datalad/git/datalad/distribution/get.py", line 501, in _install_targetpath
    ds, target_path, reckless, refds_path, description=description):
  File "/home/mih/hacking/datalad/git/datalad/distribution/get.py", line 383, in _install_necessary_subdatasets
    description=description):
  File "/home/mih/hacking/datalad/git/datalad/distribution/get.py", line 302, in _install_subds_from_flexible_source
    **kwargs):
  File "/home/mih/hacking/datalad/git/datalad/core/distributed/clone.py", line 528, in clone_dataset
    description)
  File "/home/mih/hacking/datalad/git/datalad/core/distributed/clone.py", line 754, in postclonecfg_annexdataset
    repo._init(description=description)
  File "/home/mih/hacking/datalad/git/datalad/support/annexrepo.py", line 1274, in _init
    annex_options=opts)
  File "/home/mih/hacking/datalad/git/datalad/support/annexrepo.py", line 1097, in _run_annex_command
    **kwargs)
  File "/home/mih/hacking/datalad/git/datalad/cmd.py", line 143, in run_gitcommand_on_file_list_chunks
    results.append(func(cmd, *args, **kwargs))
  File "/home/mih/hacking/datalad/git/datalad/cmd.py", line 504, in run
    **results,
datalad.support.exceptions.CommandError: CommandError: 'git-annex init -c annex.dotfiles=true -c annex.retry=3' failed with exitcode 1 under /media/mih/Samsung_T5/tmp/datalad_temp_test_clone_unborn_head_subh2nbemzo/b/sub
```

**Comment by mih** [comment]
When I go in and rerun `git annex init` after that failure I see this:

```
(datalad3-dev) mih@meiner ...t_clone_unborn_head_subj2v35m_r/b/sub (git)-[adjusted/master(unlocked)] %git annex init
init  
  Detected a filesystem without fifo support.

  Disabling ssh connection caching.

  Detected a crippled filesystem.
(scanning for unlocked files...)
fatal: refs/remotes/origin/master: not a valid SHA1

git-annex: git [Param "update-ref",Param "refs/heads/master",Param "refs/remotes/origin/master"] failed
CallStack (from HasCallStack):
  error, called at ./Git/Command.hs:42:17 in main:Git.Command
failed
git-annex: init: 1 failed
```

@kyleam you contributed that test. What should happen in this case?

**Comment by kyleam** [comment]
The repo should end up on an adjusted branch for `other`.

I suspect there is more going on here than "this was always broken".  gh-4407 specifically dealt with adjusted branch compatibility.  It points to a passing crippledfs run where this test passed.  (Checking the build again, that seems to be the case.)  Plus, I can now trigger this failure with `tools/eval_under_testloopfs`, and I'm pretty sure I used that when developing that PR, so it seems likely it was passing then.

I'll look into this more.

**Comment by bpoldrack** [comment]
@kyleam considering PR #4733 and/or PR #4708 it might well be, that it "passed" simply because the tests didn't actually use the FS they are supposed to use.

**Comment by kyleam** [comment]
> considering PR #4733 and/or PR #4708 it might well be, that it "passed" simply because the tests didn't actually use the FS they are supposed to use.

Thanks @bpoldrack.  That is certainly worth keeping in mind when interpreting the result of that github CI run.  So, it could be that that's why the CI run passed, and I'm incorrect in my guess that I ran `tools/eval_under_testloopfs` locally back then.  (In other words, I could have triggered this error back then as I can now if I had tried.)

But it turns out something did change here.  This doesn't fail until Git's 7c5045fc18 (ref-filter: apply fallback refname sort only after all user sorts, 2020-05-03), which was released with v2.27.0.  (Of course it'd be good if @mih could confirm that he is using that Git version or later.)

I actually dealt with this git commit before in b91791046 (TST: clone: Fix clock-dependent test, 2020-06-17).  As was the case there, I very much doubt that change itself is bad, but it is exposing logic bugs elsewhere.  Next step is to figure out where that elsewhere is (our end, a for-each-ref consumer in git, or git-annex).
